### PR TITLE
UPSTREAM: 15914: make kubelet image pulls serialized by default

### PR DESCRIPTION
**Pull request by derekwaynecarr**
/cc @smarterclayton @ncdc @eparis @liggitt

### CODE REVIEW

File: Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go [hunk 1]
```diff
@@ -148,6 +148,9 @@ type KubeletServer struct {
 
 	KubeApiQps   float32
 	KubeApiBurst int
+
+	// Pull images one at a time.
+	SerializeImagePulls bool
```
Comment (pmorie) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: Isn't this going to be false by default now?  I don't see where the default value is set to true.

File: Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go [hunk 2]
```diff
@@ -681,6 +687,7 @@ func SimpleKubelet(client *client.Client,
 		ResolverConfig:      kubelet.ResolvConfDefault,
 		ResourceContainer:   "/kubelet",
 		RootDirectory:       rootDir,
+		SerializeImagePulls: true,
```
Comment (eparis) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: @pmorie here?

File: Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go (hunk 1 above)
Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: Needs to go into the release notes, and a description of how to turn it off (example of config change)

File: Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go [hunk 3]
```diff
@@ -187,6 +190,8 @@ func NewKubeletServer() *KubeletServer {
 		HTTPCheckFrequency:          20 * time.Second,
 		ImageGCHighThresholdPercent: 90,
 		ImageGCLowThresholdPercent:  80,
+		KubeApiQps:                  5.0,
```
Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: Is this new?

Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: Shouldn't this be in another PR?

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/container/serialized_image_puller.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+// NewSerializedImagePuller takes an event recorder and container runtime to create a
+// image puller that wraps the container runtime's PullImage interface.
+// Pulls one image at a time.
+// Issue #10959 has the rationale behind serializing image pulls.
+func NewSerializedImagePuller(recorder record.EventRecorder, runtime Runtime, imageBackOff *util.Backoff) ImagePuller {
+	imagePuller := &serializedImagePuller{
+		recorder:     recorder,
+		runtime:      runtime,
+		backOff:      imageBackOff,
+		pullRequests: make(chan *imagePullRequest, 10),
```
Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/container/serialized_image_puller.go]: Does this need to be parameterized?

File: Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go [hunk 5]
```diff
@@ -210,9 +216,6 @@ func NewKubeletServer() *KubeletServer {
 		RootDirectory:       defaultRootDir,
 		SyncFrequency:       10 * time.Second,
 		SystemContainer:     "",
-		ReconcileCIDR:       true,
-		KubeApiQps:          5.0,
```
Comment (eparis) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: @clayton came from here, likely a gofmt thing? maybe? seems like a useless change.

File: Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go (hunk 3 above)
Comment (eparis) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: it's an alphabetization of initialization....

File: Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go (hunk 5 above)
Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: Ok, that's fine.

File: Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go (hunk 3 above)
Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: Ok

**Comment by smarterclayton** [comment]
A few comments, otherwise looks ok.

File: Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/container/serialized_image_puller.go (hunk 4 above)
Comment (smarterclayton) [inline Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubelet/container/serialized_image_puller.go]: It's fine for now

**Comment by eparis** [comment]
[test]

File: Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go (hunk 2 above)
Comment (pmorie) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: @eparis That's only for the 'simple' method, which AFAIK is only used for testing now.  The default values for the CLI flag is going to be 'false', which is the opposite of what this PR says it does.

Comment (derekwaynecarr) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: @pmorie - good catch, I had missed a line.

**Comment by derekwaynecarr** [comment]
Let me take one final pass at this, I have confused myself slightly, will poke shortly.

File: Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go (hunk 2 above)
Comment (derekwaynecarr) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: ok - this was a change between 3.0.2 backport and now.

https://github.com/openshift/ose/pull/94/files#diff-b1cd32a6f03a1a0a5f5c1b0994cc3eeeR363

I will have to add a follow-on commit to enable it.

**Comment by derekwaynecarr** [comment]
Opened upstream commit for the incorrect default value:
https://github.com/kubernetes/kubernetes/pull/16749

Will add it as a separate commit to this PR.

File: Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go (hunk 1 above)
Comment (derekwaynecarr) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: fixed in last UPSTREAM commit, the upstream had been using an incorrect default.
https://github.com/kubernetes/kubernetes/pull/16749

Comment (derekwaynecarr) [inline Godeps/_workspace/src/k8s.io/kubernetes/cmd/kubelet/app/server.go]: Release note with description on how to turn off is here:
https://github.com/openshift/openshift-docs/issues/1131#issuecomment-153485270

**Comment by derekwaynecarr** [comment]
This should be good to go now.

Confirmed via manual testing that image pull is serialized by default.

Confirmed via manual testing that the default can be changed per procedure:
https://github.com/openshift/openshift-docs/issues/1131#issuecomment-153485270

**Comment by pmorie** [comment]
LGTM

**Comment by pmorie** [comment]
[test]

**Comment by eparis** [comment]
LGTM

**Comment by smarterclayton** [comment]
You need to run generations

```
21:58:42.773370935 -0500
@@ -565,6 +565,7 @@ _openshift_start_kubernetes_kubelet()
     flags+=("--rkt-stage1-image=")
     flags+=("--root-dir=")
     flags+=("--runonce")
+    flags+=("--serialize-image-pulls")
     flags+=("--stderrthreshold=")
     flags+=("--streaming-connection-idle-timeout=")
     flags+=("--sync-frequency=")
FAILURE: Generated completions out of date. Please run hack/update-generated-completions.sh
make[1]: Leaving directory `/data/src/github.com/openshift/origin'
```

**Comment by derekwaynecarr** [comment]
@smarterclayton @eparis - done

**Comment by smarterclayton** [comment]
[merge]
